Sorting algos (#9)
Ada-Developers-Academy/core-unit-2: sorting-algos/bubble-sort.md

| 15.       | [35, 40, 16, 45, 11, 7, **50**, **90**, 99] | 50 and 90 are compared, and don't swap.                                                                    |
| -         | -                                           | We do not compare 90 and 99,. In the bubble sort algorithm, the range of items to check shrinks each loop. |

We can describe even one more whole loop:

| Iteration | Array                                       |
| --------- | ------------------------------------------- | --------------------------------------------------------------------------------------------------------- |
| 16.       | [**35**, **40**, 16, 45, 11, 7, 50, 90, 99] |
| 17.       | [35, **40**, **16**, 45, 11, 7, 50, 90, 99] |
| 18.       | [35, 16, **40**, **45**, 11, 7, 50, 90, 99] |
| 19.       | [35, 16, 40, **45**, **11**, 7, 50, 90, 99] |
| 20.       | [35, 16, 40, 11, **45**, **7**, 50, 90, 99] |
| 22.       | [35, 16, 40, 11, 7, **45**, **50**, 90, 99] |
| -         | -                                           | We do not compare 50 and 90. In the bubble sort algorithm, the range of items to check shrinks each loop. |

The algorithm continues!

## Big(O) Complexity

> The time complexity of Bubble sort is _O(n<sup>2</sup>)_. We can conclude this by considering:
> 
> - The inner most swap conditional statement is _O(1)_
> - The inner loop runs _i_ times i.e. _O(i)_ time complexity
> - The outer loop runs from _i_ = _0_ to _n_
> 
> Overall, the inner loop runs _n-1_ times during the first iteration of the outer loop, _n-2_ times during the second iteration through the outer loop, and so on. i.e. _n-1 + n-2 + n-3 + ... + 3 + 2 + 1_ = _n(n-1)/2_ = _O(n<sup>2</sup>)_

Reviewer: It's been a bit since the big O unit, so I'd like to expand this significantly (as the first of the 3 algorithms) to review concepts from the previous big O lesson and how they apply here.

```suggestion
There are two basic operations in which we are interested when analyzing sorting algorithms:

1. the number of comparisons
1. the number of swaps

In our big O analysis of sorting algorithms, we will typically focus on the number of comparisons. In some languages, the number of swaps will also be important and could provided a reason to favor an algorithm with fewer swaps over an algorithm with many swaps.

Also, recall that big O analysis considers the worst case for an algorithm! Although it's possible for bubble sort to finish after a single pass (_O(n)_) if the array happened to be sorted already, we are less interested in that situation.

Noticing best case performance can be useful if we are choosing between multiple algorithms that otherwise have the same complexity, or if we have forehand knowledge about the data with which we will be working.

In the worst case, we would need to perform the check and swap with each pair, and everything would be out of order so that we don't finish early. This would require about as many passes through the array as there are values.

This gives us a time complexity for bubble sort of _O(n<sup>2</sup>)_, where _n_ is the array length.

We conclude this by considering:

- The outer loop (overall number of passes) must run _n_ times
- The inner loop (performing comparisons and swapping) must also run _n_ times

We can even be a little more precise with our counting.

The inner loop runs _n-1_ times during the first iteration of the outer loop, _n-2_ times during the second iteration through the outer loop, and so on. i.e. _n-1 + n-2 + n-3 + ... + 3 + 2 + 1_.

There is a mathematical identity which states this summation is the same as _n(n-1)/2_. When multiplied out this results in _1⁄2n<sup>2</sup>-1⁄2n_, for _O(n<sup>2</sup>)_.

Notice that _n(n-1)/2_ does _not_ equal _n<sup>2</sup>_. Recall that when performing big O analysis, we are interested in the rate of growth in the algorithm rather than the exact number of operations. The growth rate will be dominated by the largest term in the equation.

After multiplying out the expression, the largest term is _1⁄2n<sup>2</sup>_. After dropping the constant, this allows us to say that bubble sort is _O(n<sup>2</sup>)_.
```

### !callout-info

## Mathematical Proofs Out of Scope

> The mathematical proof is out of scope for our discussion. However, there are great resources online that explain the math in-depth!

Reviewer: some rewording and add FYC

```suggestion
In general, full mathematical proofs for the algorithms we discuss are out of scope for the curriculum. We will often apply analyses that are more intuitive than rigorous. However, there are great resources online that explain the math in-depth! Follow your curiosity!
```

### !end-callout

> 

Reviewer: add stability discussion (as mentioned in the sorting lessons review)

```suggestion

## Stability

Bubble sort is a stable sorting algorithm. Since we only shift values by one position, and even then only if they are strictly out of order, equivalent values will never "jump" over one another.

That is, if the value 10 appeared in the list twice, the leftmost instance would always remain to the left, while the rightmost would remain to the right.

As discussed in the Sorting Algorithms lesson, stability can be a useful property for a sorting algorithm to have. If our values were more complex records that we would like to sort based on multiple attributes, then we can simply apply a stable sort multiple times. Or if the initial ordering of a list isn't entirely random, a stable sort will preserve the ordering between records that are otherwise equivalent.

```

## Example Implementation

Consider this example implementation of bubble sort.

```python

> def bubble_sort(array):
>     i = 0
>     # Begin the outer loop
>     while i < len(array) - 1:
>         j = 0
>         # Begin the inner loop
>         while j < (len(array) - i - 1):
>             # Compare two adjacent items
>             if array[j] > array[j+1]:
>                 # Swap
>                 temp = array[j]
>                 array[j] = array[j+1]
>                 array[j + 1] = temp
>             j += 1
>         i += 1
>     return array

Reviewer: consistent spacing for `[j+1]`
remove the return since this is modifying the passed in array (the suggested explanation below mentions this explicitly)

```suggestion
def bubble_sort(array):
    i = 0
    # Begin the outer loop
    while i < len(array) - 1:
        j = 0
        # Begin the inner loop
        while j < (len(array) - i - 1):
            # Compare two adjacent items
            if array[j] > array[j+1]:
                # Swap
                temp = array[j]
                array[j] = array[j+1]
                array[j+1] = temp
            j += 1
        i += 1
```

```

> We should compare this code with our detailed explanation:

Reviewer: consistent verbiage with the other lessons

```suggestion
Compare this code with this detailed explanation of the algorithm:
```

> - Start a variable `i` at `0`
> - Create an outer loop that loops while `i` is smaller than the length of the array - 1
>   - Start a variable `j` at `0`
>   - Create an inner loop that loops while `j` is smaller than `len(array) - i - 1`
>     - Compare the items at index `j` and `j+1`. If they're out of order:
>       - Swap the items at `array[j]` and `array[j+1]` using a temporary variable, `temp`
>     - Increment `j`
>   - Increment `i`

Reviewer: expand the explanation, both "inline" as well as general observations

```suggestion
- Start a variable `i` at `0`
- Create an outer loop that loops while `i` is smaller than the length of the `array - 1` since we don't need to bubble the final value
  - Start a variable `j` at `0`
  - Create an inner loop that loops while `j` is smaller than `len(array) - i - 1` since we want to avoid the `i` sorted elements at the end, and we need to leave one extra value for the pairwise comparisons
    - Compare the items at index `j` and `j+1`. If they're out of order:
      - Swap the items at `array[j]` and `array[j+1]` using a temporary variable, `temp`
    - Increment `j`
  - Repeat until `j` reaches its limit
  - Increment `i`
- Repeat until `i` reaches its limit

This implementation does _not_ explicitly track whether a swap has been made (which could allow early termination). But when we get down to a single item in the unsorted portion, this is implicitly true, as there is nothing left with which we can compare.

This implementation sorts the array in-place. That is, it uses only a small amount of constant extra memory (just a few local variables), and the original array itself is updated. Because the original array is modified, no return statement is needed.

Note that the presented implementation used a temporary variable to perform the swap. This is a common pattern in languages that don't support multiple values in a single assignment. This implementation has been written in a general manner to emphasize that the algorithm—the steps for performing bubble sort—are not specific to any one language.

We will follow this same style when discussing other sorting algorithms. But we should feel free to modify the presented implementation to make better use of language-specific features!
```
